gammapy/utils/fitting/datasets.py

```diff
@@ -141,3 +141,90 @@ def __str__(self):
     def copy(self):
         """A deep copy."""
         return copy.deepcopy(self)
+
+    def set_models_from_yaml(self, filemodel):
+        """add models and backgrounds to datasets from yaml files
+    
+        Parameters
+        ----------
+        fdatasets : str
+            filepath to yaml datasets file
+        filemodel : str
+            filepath to yaml models file
+        """
+        from ..serialization import models_to_datasets
+        from ..scripts import read_yaml
+
+        components = read_yaml(filemodel)
+        models_to_datasets(self, components, get_lists=False)
+
+    @classmethod
+    def from_yaml(cls, filedata, filemodel=None):
+        from ...cube.fit import MapDataset
+        from ..scripts import read_yaml
+
+        """De-serialize datasets from yaml and fits files
+    
+        Parameters
+        ----------
+        fdatasets : str
+            filepath to yaml datasets file
+        filemodel : str
+            filepath to yaml models file
+            
+        Returns
+        -------
+        dataset : `Datasets`
+            'gammapy.utils.fitting.Datasets'
+         """
+
+        data_list = read_yaml(filedata)
+        datasets_list = []
+        for data in data_list["datasets"]:
+            dataset = MapDataset.read(data["filename"])
+            dataset.obs_id = data["id"]
+            datasets_list.append(dataset)
+        datasets = cls(datasets_list)
+        if filemodel is not None:
+            datasets.set_models_from_yaml(filemodel)
+        return datasets
+
+    def to_yaml(self, path, selection="all", overwrite=False):
+        """Serialize datasets to yaml and fits files
+
+        Parameters
+        ----------
+        path : str
+            path to write files
+        selection : {"all", "simple"}
+            "simple" option reduce models parameters attributes displayed to only 
+            name, value, unit,frozen
+        """
+        from ..serialization import models_to_dict
+        from ..scripts import write_yaml
+        from ...cube.models import BackgroundModels, SkyModels
+
+        models_list = []
+        backgrounds_list = []
+        datasets_dictlist = []
+        for dataset in self.datasets:
+            filename = str(path) + "maps_" + dataset.obs_id + ".fits"
+            dataset.write(filename, overwrite)
+            datasets_dictlist.append({"id": dataset.obs_id, "filename": filename})
+
+            if isinstance(dataset.background_model, BackgroundModels):
```

Comment:
I agree serialization would be simpler if datasets were initialized with only lists.

Is there something equivalent to SkyModels and BackgroundModels for Spectral models ?

I guess removing all the instance checks and force to give only lists as you suggest  will break several tests and notebooks, I would prefer to look at this in a separate pull-request.

Reply:
Then leave the TODO for now, and we can continue the discussion in  #2102 .
Would be great if you take that on in a separate PR.
The way forward is to just simplify the code, see what breaks, and adjust tests and examples if needed.

It can be OK to accept several things on input. But then before storing it as a data member, one should always go do a uniform representation. This way, when accessing data members like you do here, you no longer have to deal with the multiple cases, that only happens once in the setter.

Reply:
Doing test-driven development really helps to write such code.
E.g. the case where you have `BackgroundModels[...]` that's broken should just be deleted, and only be added when you have a use and test case that requires it.